File: ai/src/scheduler/cron.ts
/**
 * F7 — Cron Scheduler for Continuous Prediction Updates
 * Uses dynamic require to resolve node-cron from backend's node_modules.
 */
import * as path from 'path';
const cron = require(require.resolve('node-cron', {
    paths: [path.resolve(__dirname, '../../../backend/node_modules')]
}));

import { computeRiskScores } from '../services/risk.service';
import { generateGuideAlerts, generateCoordinatorAlerts } from '../services/alerts.service';
import { batchCheckApprovals } from '../services/approval.service';
import { getCronIntervals, loadConfig } from '../engine/thresholds';

let scheduledJobs: Record<string, any> = {};

export const initScheduler = async () => {
    await loadConfig();
    const intervals = getCronIntervals();

    // Risk scoring — every 6 hours by default
    scheduledJobs['risk_scoring'] = cron.schedule(intervals.risk_scoring, async () => {
        console.log('🤖 [CRON] Running risk scoring...');
        try {
            const count = await computeRiskScores();
            console.log(`🤖 [CRON] Risk scoring complete — ${count} projects scored`);
        } catch (err) { console.error('🤖 [CRON] Risk scoring failed:', err); }
    });

    // Alert generation — daily at 8 AM by default
    scheduledJobs['alerts'] = cron.schedule(intervals.alerts, async () => {
        console.log('🤖 [CRON] Generating alerts...');
        try {
            const guideCount = await generateGuideAlerts();
            const coordCount = await generateCoordinatorAlerts();
            console.log(`🤖 [CRON] Alerts generated — guide: ${guideCount}, coordinator: ${coordCount}`);
        } catch (err) { console.error('🤖 [CRON] Alert generation failed:', err); }
    });

    // Auto-approval check — daily at midnight by default
    scheduledJobs['predictions'] = cron.schedule(intervals.predictions, async () => {
        console.log('🤖 [CRON] Checking auto-approvals...');
        try {
            const result = await batchCheckApprovals();
            console.log(`🤖 [CRON] Approval check — ${result.eligible.length} eligible of ${result.total_checked}`);
        } catch (err) { console.error('🤖 [CRON] Approval check failed:', err); }
    });

    console.log('🤖 AI Scheduler initialized with intervals:', intervals);
};

/**
 * Manually trigger a specific job
 */
export const triggerJob = async (jobName: string): Promise<string> => {
    switch (jobName) {
        case 'risk_scoring': {
            const count = await computeRiskScores();
            return `Risk scoring complete — ${count} projects scored`;
        }
        case 'alerts': {
            const g = await generateGuideAlerts();
            const c = await generateCoordinatorAlerts();
            return `Alerts generated — guide: ${g}, coordinator: ${c}`;
        }
        case 'predictions': {
            const r = await batchCheckApprovals();
            return `Approval check — ${r.eligible.length} eligible of ${r.total_checked}`;
        }
        default:
            throw new Error(`Unknown job: ${jobName}`);
    }
};

export const stopScheduler = () => {
    for (const [name, job] of Object.entries(scheduledJobs)) {
        (job as any).stop();
        console.log(`🤖 Stopped cron job: ${name}`);
    }
    scheduledJobs = {};
};

File: ai-py/scheduler/cron.py
"""
F7 — Continuous Prediction Updates via APScheduler
"""
from apscheduler.schedulers.background import BackgroundScheduler
from services.risk_service import compute_risk_scores
from services.alerts_service import generate_guide_alerts, generate_coordinator_alerts
from services.approval_service import batch_check_approvals

_scheduler: BackgroundScheduler | None = None


def _run_risk_scoring():
    print("🤖 [CRON] Running risk scoring...")
    try:
        count = compute_risk_scores()
        print(f"🤖 [CRON] Risk scoring complete — {count} projects scored")
    except Exception as e:
        print(f"🤖 [CRON] Risk scoring failed: {e}")


def _run_alerts():
    print("🤖 [CRON] Generating alerts...")
    try:
        g = generate_guide_alerts()
        c = generate_coordinator_alerts()
        print(f"🤖 [CRON] Alerts generated — guide: {g}, coordinator: {c}")
    except Exception as e:
        print(f"🤖 [CRON] Alert generation failed: {e}")


def _run_approvals():
    print("🤖 [CRON] Checking auto-approvals...")
    try:
        result = batch_check_approvals()
        print(f"🤖 [CRON] Approval check — {len(result['eligible'])} eligible of {result['total_checked']}")
    except Exception as e:
        print(f"🤖 [CRON] Approval check failed: {e}")


def init_scheduler():
    global _scheduler
    _scheduler = BackgroundScheduler()

    # Risk scoring — every 6 hours
    _scheduler.add_job(_run_risk_scoring, "interval", hours=6, id="risk_scoring")

    # Alert generation — daily at 8 AM
    _scheduler.add_job(_run_alerts, "cron", hour=8, id="alerts")

    # Auto-approval check — daily at midnight
    _scheduler.add_job(_run_approvals, "cron", hour=0, id="predictions")

    _scheduler.start()
    print("🤖 AI Scheduler initialized (APScheduler)")


def trigger_job(job_name: str) -> str:
    if job_name == "risk_scoring":
        count = compute_risk_scores()
        return f"Risk scoring complete — {count} projects scored"
    elif job_name == "alerts":
        g = generate_guide_alerts()
        c = generate_coordinator_alerts()
        return f"Alerts generated — guide: {g}, coordinator: {c}"
    elif job_name == "predictions":
        r = batch_check_approvals()
        return f"Approval check — {len(r['eligible'])} eligible of {r['total_checked']}"
    else:
        raise ValueError(f"Unknown job: {job_name}")


def stop_scheduler():
    global _scheduler
    if _scheduler:
        _scheduler.shutdown()
        print("🤖 Scheduler stopped")
        _scheduler = None
